pyccel/ast/cupyext.py

```diff
@@ -436,6 +444,39 @@ def __str__(self):
 
 #==============================================================================
 
+class CupyRavel(CupyArray):
+    """
+    Class representing a call to the cupy ravel function which
+    returns flattened version of the passed array
+
+    Parameters
+    ==========
+    arg : PyccelAstNode
+            A PyccelAstNode of unknown shape
+    memory_location : string
+            The location where the new array memory should be allocated
+    """
+    name   = 'ravel'
+    __slots__ = ()
+    _attribute_nodes = ('_arg',)
+    def __new__(cls, arg):
+        if not isinstance(arg, (list, tuple, PyccelAstNode)):
+            raise TypeError('Unknown type of  %s.' % type(arg))
+        if arg.rank == 0:
+            raise TypeError('Unknown type of  %s.' % type(arg))
+        return super().__new__(cls)
+
+    def __init__(self, arg):
+        super().__init__(arg = arg)
+        shape = reduce((lambda x, y: x.python_value * y.python_value), self.shape)
+        self._shape = [shape if isinstance(shape, LiteralInteger) else LiteralInteger(shape)]
+        self._rank = 1
```

Comment:
If this is always 1 can it be a class attribute? Or is it here due to the inheritance from `CupyArray`?

Reply:
We need that `_rank` to be inherited from `CupyArray` since we will use the normal `CupyArray` printer in some cases.

Reply:
Yeah we can't really put its as class attribute since we inherit from `CupyArray` and we use `CupyArray`'s printer in some cases.

Reply:
Ok. So `CupyRavel` is a special case of `CupyArray`? I'm surprised using `CupyArray`'s printer doesn't cause problems with the `arg` property. I would have expected `Ravel` to be more or less equivalent to `_flatten_list`

Reply:
In some cases we will need to copy the array that's where we use `copy_CudaArray_Data` like in `CupyArray`, but in some other cases we need to create a view instead.

Reply:
If it can be a view then how can it be a special case of `CupyArray`? I am not convinced by this inheritance

Reply:
It's just that if the original array isn't a `device` array, it will create a copy the same way we handle it in a regular `CupyArray`, what do you suggest instead ?

Reply:
Hmm, I'm not sure. Maybe `CupyRavel` to act as the reshape, but if `memory_location != 'device'` then do `self._arg = CupyArray(arg)` ? But that will create an unnecessary temporary. Can `CupyArray` take a ravel object as an argument?

I'm almost tempted to say that it should be something like:
```python
class CupyRavel:
    def __new__(cls, arg):
        if arg.memory_location == 'device':
            return Reshape(arg)
        else:
            return Reshape(CupyArray(arg))
```

Do cupy's arrays have a `reshape` attribute like numpy's?

Reply:
I think its a good idea, but it would be better for it to have it own PR.

Reply:
Maybe, I definitely agree that `Reshape` should have its own PR,